Out of curiosity, I tried to use this fix to get the derivative of Flax CLIP exporting correctly, and before you added your asserts back in it failed on the unique index assertion:
```
AssertionError: Indices must be unique for this implementation. For other cases look at using something with tf.math.unsorted_segment_*
```
But with your asserts added back in I get this error:
```
    854     tf.assert_equal(tf.rank(operand), tf.rank(updates)+1, message='Updates must be 1 dimension less than operand')
    855     tf.assert_rank(scatter_indices, 1, message='Scatter indices must be 1 dimensional')
--> 856     assert (len(wd)==len(sd)==1) & (wd==sd) & (ud==tuple(range(max(ud)+1))), 'Unsupported scatter op'
    857     tf.assert_rank(operand, len(ud)+1, message='Unsupported scatter op')
    858     axis = wd[0] # axis the `scatter_indices` index into within `operand`

ValueError: max() arg is an empty sequence
```
So it looks like an extra assert might be needed above that line? Here's a minimal notebook: https://colab.research.google.com/drive/1SPeT3bBBSzZpAz0kqgfzvvYCsTXIPJIb

(It would be really cool if this PR added support for CLIP too, in case it's not too hard to support non-unique indices, but you've already been absurdly helpful so please ignore this if you don't have the bandwidth right now to do this!)

--- a/jax/experimental/jax2tf/g3doc/no_xla_limitations.md
+++ b/jax/experimental/jax2tf/g3doc/no_xla_limitations.md
@@ -33,7 +33,7 @@ For a detailed description of these XLA ops, please see the
 | XlaConv | `lax.conv_general_dilated` | [Partial](#xlaconv) |
 | XlaGather | `lax.gather` | [Partial](#xlagather) |
 | XlaReduceWindow | `lax.reduce_window_sum_p`, `lax.reduce_window_min_p`, `lax.reduce_window_max_p`, and `lax.reduce_window_p` | [Partial](#xlareducewindow) |
-| XlaScatter | `lax.scatter_p`, `lax.scatter_min_p`, `lax.scatter_max_p`, `lax.scatter_mul_p`, `lax.scatter_add_p` | Unsupported |
+| XlaScatter | `lax.scatter_p`, `lax.scatter_min_p`, `lax.scatter_max_p`, `lax.scatter_mul_p`, `lax.scatter_add_p` | [Partial](#xlascatter) |
 | XlaSelectAndScatter | `lax.select_and_scatter_add_p` | Unsupported |
 | XlaReduce | `lax.reduce`, `lax.argmin`, `lax.argmax` | Unsupported |
 | XlaVariadicSort | `lax.sort` | Unsupported |
@@ -153,3 +153,15 @@ We support these ops with the following limitations:
 * `padding` should either be `VALID` or `SAME`.
 
 `lax.reduce_window_min_p` and `lax.reduce_window` are currently not supported.
+
+### XlaScatter
+
+This op is called by `lax.scatter`, `lax.scatter_min`, `lax.scatter_max`, 
+`lax.scatter_mul` and `lax.scatter_add`. 
+
+We support all these ops for unique indices. For non-unique indices we 
+support (min,max,mul,add) for single depth scatters.
+
+There are a few more limitations:
+* the GatherScatterMode must be PROMISE_IN_BOUNDS.
+* dtypes `np.bool` and `jnp.complex*` are not supported.

--- a/jax/experimental/jax2tf/impl_no_xla.py
+++ b/jax/experimental/jax2tf/impl_no_xla.py
@@ -792,10 +792,128 @@ def _dynamic_update_slice(operand, update, *start_indices,
 
 tf_impl_no_xla[lax.dynamic_update_slice_p] = _dynamic_update_slice
 
-tf_impl_no_xla[lax.scatter_p] = _unimplemented("scatter")
-tf_impl_no_xla[lax.scatter_min_p] = _unimplemented("scatter min")
-tf_impl_no_xla[lax.scatter_max_p] = _unimplemented("scatter max")
-tf_impl_no_xla[lax.scatter_mul_p] = _unimplemented("scatter mul")
-tf_impl_no_xla[lax.scatter_add_p] = _unimplemented("scatter add")
+
+def shift_axes_forward(operand, axes: tuple, inverse: bool=False,
+                       forward: bool=True):
+  """Shifts the tuple of axes to the front of an array"""
+  other_axes = tuple([i for i in range(len(operand.shape)) if i not in axes])
+  fwd_order = axes + other_axes if forward else other_axes + axes
+  order = fwd_order if not inverse else tf.math.invert_permutation(fwd_order)
+  return tf.transpose(operand, order)
+
+def convert_scatter_jax_to_tf(update_op, unsorted_segment_op=None):
+  def error(msg):
+    suffix = ("See source code for the precise conditions under which "
+              "scatter_(update/add/multiply/min/max) ops can be converted without XLA.")
+    return _xla_disabled_error("scatter_(update/add/multiply/min/max)", f"{msg} - {suffix}")
+
+  def _sparse_scatter(
+    operand,
+    scatter_indices,
+    updates,
+    update_jaxpr,
+    update_consts,
+    dimension_numbers,
+    indices_are_sorted: bool,
+    unique_indices: bool,
+    mode,
+    _in_avals: Sequence[core.ShapedArray],
+    _out_aval: core.ShapedArray):
+    """
+    Implementation of scatter specialised to indexing from the
+    front axes. This covers unique indices and non-unique indices
+    of single depth.
+
+    Note on unique indices: `tf.tensor_scatter_nd_update` interprets
+    indices thusly: every axis except the final one encodes a batch
+    dimension, the final axis encoding the actual indices to scatter in to.
+    It enforces, at least one, batch dimension so we add an empty
+    dimension to indices and updates if lacking.
+
+    Note on non-unique indices: There is no tf op for non single depth
+    indexing. But if indexing is single depth, this can be viewed as a
+    segment op.
+    """
+    # Infer unique indices from lack of batch dimension
+    unique_indices = unique_indices or (len(scatter_indices.shape) == 1)
+    if unique_indices:
+      suboperand = tf.gather_nd(operand, scatter_indices)
+      updated_suboperand = update_op(suboperand, updates)
+      # add a batch dim if none exist
+      if len(scatter_indices.shape) == 1:
+        scatter_indices = scatter_indices[None]
+        updated_suboperand = updated_suboperand[None]
+      y = tf.tensor_scatter_nd_update(operand, scatter_indices, updated_suboperand)
+    else:
+      if (scatter_indices.shape[-1] == 1) and (unsorted_segment_op != None):
+        # If only indexing into the first dimension, it's a segment op
+        operand_update = unsorted_segment_op(updates, tf.squeeze(scatter_indices, -1), operand.shape[0])
+        y = update_op(operand, operand_update)
+      else:
+        raise error("Scatter supports unique indices. Scatter also supports non-unique indices with indexing into only one dimension for (add, mul, min, max)")
+    return y
+
+  def sparse_scatter(
+    operand,
+    scatter_indices,
+    updates,
+    update_jaxpr,
+    update_consts,
+    dimension_numbers,
+    indices_are_sorted: bool,
+    unique_indices: bool,
+    mode,
+    _in_avals: Sequence[core.ShapedArray],
+    _out_aval: core.ShapedArray):
+    """
+    Wrapper around the scatter function.
+    The underlying tf ops `tf.tensor_scatter_nd_update` and
+    `tf.math.unsorted_segment_*` index from the front dimensions.
+    `tf.math.unsorted_segment_*` indexs to a depth 1 from the front.
+    `tf.tensor_scatter_nd_update` indexs from the front dimensions onwards
+    , with no ability to skip a dimension. This function
+    shifts the axes to be indexed to the front then calls a front-specific
+    implementation, then inverse-shifts the output.
+
+    scatter_dims_to_operand_dims: dimensions which the scatter indexes in to.
+      We shift these to the front to match tf syntax. All other dims are batch
+    update_window_dims: dimensions which are not batch dimensions. We shift
+      these to the back as the remaining dimensions are batch dimensions.
+    """
+    ud = dimension_numbers.update_window_dims
+    wd = dimension_numbers.inserted_window_dims
+    sd = dimension_numbers.scatter_dims_to_operand_dims
+    dtype = operand.dtype # assume updates has same dtype as operand
+    if dtype in [tf.bool, tf.complex64]:
+      raise error(f"Scatter does not support operands of type {dtype}")
+    if not (wd == sd):
+      raise error("Complex scatters are not supported")
+    if not (mode == lax.GatherScatterMode.PROMISE_IN_BOUNDS):
+      raise error("Only scatter mode `PROMISE_IN_BOUNDS` is supported")
+    # Shift axes to the front to match tf syntax, inverse afterwards
+    fwd = partial(shift_axes_forward, axes=sd)
+    inv = partial(fwd, inverse=True)
+    # shift update value axes to the back, so batch are at the front
+    updates_shifted = shift_axes_forward(updates, axes=ud, forward=False)
+    return inv(_sparse_scatter(
+      fwd(operand),
+      scatter_indices,
+      updates_shifted,
+      update_jaxpr,
+      update_consts,
+      dimension_numbers,
+      indices_are_sorted,
+      unique_indices,
+      mode,
+      _in_avals,
+      _out_aval,
+    ))
+  return sparse_scatter
+
+tf_impl_no_xla[lax.scatter_p] = convert_scatter_jax_to_tf(lambda x,y: y) # just replace with the update
+tf_impl_no_xla[lax.scatter_add_p] = convert_scatter_jax_to_tf(tf.add,      tf.math.unsorted_segment_sum)
+tf_impl_no_xla[lax.scatter_mul_p] = convert_scatter_jax_to_tf(tf.multiply, tf.math.unsorted_segment_prod)
+tf_impl_no_xla[lax.scatter_min_p] = convert_scatter_jax_to_tf(tf.minimum,  tf.math.unsorted_segment_min)
+tf_impl_no_xla[lax.scatter_max_p] = convert_scatter_jax_to_tf(tf.maximum,  tf.math.unsorted_segment_max)
 
 tf_impl_no_xla[lax.sort_p] = _unimplemented("sort")

--- a/jax/experimental/jax2tf/tests/jax2tf_limitations.py
+++ b/jax/experimental/jax2tf/tests/jax2tf_limitations.py
@@ -950,6 +950,10 @@ class Jax2TfLimitation(primitive_harness.Limitation):
     return []
 
   @classmethod
+  def scatter(cls, harness):
+    return []
+
+  @classmethod
   def select_and_gather_add(cls, harness):
     return [
         # This JAX primitives is not not exposed directly in the JAX API

--- a/jax/experimental/jax2tf/tests/primitive_harness.py
+++ b/jax/experimental/jax2tf/tests/primitive_harness.py
@@ -1272,7 +1272,7 @@ def _make_scatter_harness(name,
 # Validate dtypes
 for dtype in jtu.dtypes.all:
   for f_lax in [
-      lax.scatter_add, lax.scatter_mul, lax.scatter_max, lax.scatter_min
+      lax.scatter, lax.scatter_add, lax.scatter_mul, lax.scatter_max, lax.scatter_min
   ]:
     _make_scatter_harness("dtypes", dtype=dtype, f_lax=f_lax)
 

--- a/jax/experimental/jax2tf/tests/primitive_harness.py
+++ b/jax/experimental/jax2tf/tests/primitive_harness.py
@@ -1236,11 +1236,12 @@ def _make_scatter_harness(name,
                           update_shape=(2,),
                           mode=lax.GatherScatterMode.FILL_OR_DROP,
                           dtype=np.float32,
-                          dimension_numbers=((), (0,), (0,))):
+                          dimension_numbers=((), (0,), (0,)),
+                          enable_xla=True):
   dimension_numbers = lax.ScatterDimensionNumbers(*dimension_numbers)
   define(
       f_lax.__name__,
-      f"{name}_shape={jtu.format_shape_dtype_string(shape, dtype)}_scatterindices={scatter_indices.tolist()}_updateshape={update_shape}_updatewindowdims={dimension_numbers.update_window_dims}_insertedwindowdims={dimension_numbers.inserted_window_dims}_scatterdimstooperanddims={dimension_numbers.scatter_dims_to_operand_dims}_indicesaresorted={indices_are_sorted}_uniqueindices={unique_indices}_mode={mode}"
+      f"{name}_shape={jtu.format_shape_dtype_string(shape, dtype)}_scatterindices={scatter_indices.tolist()}_updateshape={update_shape}_updatewindowdims={dimension_numbers.update_window_dims}_insertedwindowdims={dimension_numbers.inserted_window_dims}_scatterdimstooperanddims={dimension_numbers.scatter_dims_to_operand_dims}_indicesaresorted={indices_are_sorted}_uniqueindices={unique_indices}_mode={mode}_enablexla={enable_xla}"
       .replace(" ", ""),
       partial(
           f_lax,
@@ -1266,7 +1267,8 @@ def _make_scatter_harness(name,
       dimension_numbers=dimension_numbers,
       indices_are_sorted=indices_are_sorted,
       unique_indices=unique_indices,
-      mode=mode)
+      mode=mode,
+      enable_xla=enable_xla)
 
 
 # Validate dtypes
@@ -1315,6 +1317,60 @@ for mode in (lax.GatherScatterMode.PROMISE_IN_BOUNDS,
                         scatter_indices=np.array([[4], [77]]),
                         update_shape=(2,))
 
+# Validate no XLA scatters
+for dtype in set(jtu.dtypes.all) - set(jtu.dtypes.complex) - set(jtu.dtypes.boolean):
+  for f_lax in [
+      lax.scatter_add, lax.scatter_mul, lax.scatter_max, lax.scatter_min, lax.scatter
+  ]:
+    for shape, scatter_indices, update_shape, dimension_numbers in [
+        ((1,), (0,), (), ((), (0,), (0,))), # zero case
+        ((1, 1), (0,), (1,), ((0,), (0,), (0,))),
+        ((1, 1, 1), (0,), (1, 1), ((0, 1), (0,), (0,))),
+        ((1, 50, 3), (32,), (1, 3), ((0, 1), (1,), (1,))),
+        ((1, 2, 3), [1], (1, 3), ((0, 1), (1,), (1,))), # slice 2nd dim
+        ((1, 2, 3), [0], (2, 3), ((0, 1), (0,), (0,))), # slice 1st dim
+        ((1, 2, 3), [1, 2], (1,), ((0,), (1, 2), (1, 2))), # 2nd and 3rd
+        ((4, 2, 3), [3, 2], (2,), ((0,), (0, 2), (0, 2))), # 1st and 3rd
+        ((4, 2, 3, 5), [0, 4], (4, 3), ((0, 1), (1, 3), (1, 3))), # 2nd and 4th
+        ((5, 6, 7), [[0, 1],[2, 3]], (2, 7), ((1,), (0, 1), (0, 1))), # .at[((3,4),(5,5))] shapes
+        ((5, 6, 7), [[[0],[1]],[[2],[3]]], (5, 2, 2, 7), ((0, 3), (1,), (1,))), # .at[:, ((3,4),(5,5))] shapes
+        ((5, 6, 7), [[[0, 1],[2, 3]],[[4, 0],[1, 2]]], (5,2,2), ((0,), (1, 2), (1, 2))), # .at[:, ((3,4),(5,5)), 3] shapes
+    ]:
+      _make_scatter_harness("no_xla_unique_indices",
+                          shape=shape,
+                          f_lax=f_lax,
+                          unique_indices=True,
+                          scatter_indices=np.array(scatter_indices),
+                          update_shape=update_shape,
+                          mode=lax.GatherScatterMode.PROMISE_IN_BOUNDS,
+                          dtype=dtype,
+                          dimension_numbers=dimension_numbers,
+                          enable_xla=False,
+                          )
+
+# Validate no XLA scatters with non-unique indices with an indexing depth of 1
+for dtype in set(jtu.dtypes.all) - set(jtu.dtypes.complex) - set(jtu.dtypes.boolean):
+  for f_lax in [
+      lax.scatter_add, lax.scatter_mul, lax.scatter_max, lax.scatter_min
+  ]:
+    for shape, scatter_indices, update_shape, dimension_numbers in [
+        ((1,), [[0],[0]], (2,), ((), (0,), (0,))), # .at[((0,0),)]
+        ((3,), [[1],[0],[1]], (3,), ((), (0,), (0,))), # .at[((1,0,1),)]
+        ((2, 3), [[[2],[2],[2]]], (2, 1, 3), ((0,), (1,), (1,))), # 2nd dim, .at[:, ((2,2,2),)]
+        ((3, 5, 40), [[1],[1]], (3, 5, 2), ((0, 1), (2,), (2,))),
+        ((3, 5, 4), [[1],[1]], (3, 2, 4), ((0, 2), (1,), (1,))),
+    ]:
+      _make_scatter_harness("no_xla_non_unique_indices",
+                          shape=shape,
+                          f_lax=f_lax,
+                          unique_indices=False,
+                          scatter_indices=np.array(scatter_indices),
+                          update_shape=update_shape,
+                          mode=lax.GatherScatterMode.PROMISE_IN_BOUNDS,
+                          dtype=dtype,
+                          dimension_numbers=dimension_numbers,
+                          enable_xla=False,
+                          )
 
 for dtype in jtu.dtypes.all:
   arg_shape = (2, 3)

